packages/@ourworldindata/core-table/src/CoreTable.ts

```diff
@@ -1441,15 +1440,65 @@ export class CoreTable<
      *   ```
      *
      */
-    complete(columnSlugs: ColumnSlug[]): this {
-        const index = this.rowIndex(columnSlugs)
-        const cols = this.getColumns(columnSlugs)
-        const product = cartesianProduct(...cols.map((col) => col.uniqValues))
-        const toAdd = product.filter((row) => !index.has(row.join(" ")))
-        return this.appendRows(
-            rowsFromMatrix([columnSlugs, ...toAdd]),
+    complete(columnSlugs: [ColumnSlug, ColumnSlug]): this {
+        if (columnSlugs.length !== 2)
+            throw new Error("Can only run complete() for exactly 2 columns")
+
+        console.time("c")
+        const [slug1, slug2] = columnSlugs
+        const col1 = this.get(slug1)
+        const col2 = this.get(slug2)
+
+        console.timeLog("c", "got cols")
+
+        const cartesianProductSize =
+            col1.uniqValues.length * col2.uniqValues.length
+        if (this.numRows >= cartesianProductSize) {
+            if (this.numRows > cartesianProductSize)
+                throw new Error("Table has more rows than expected")
+
+            console.timeLog("c", "skip", cartesianProductSize, this.numRows)
+            console.timeEnd("c")
+            // Table is already complete
+            return this
+        }
+
+        // Map that points from a value in col1 to a set of values in col2
+        // At first we fill it with the cartesian product of unique values in col1 and col2,
+        // then we remove all the values that already exist in the table.
+        const missingValues = new Map<CoreValueType, Set<CoreValueType>>()
+        for (const val1 of col1.uniqValues) {
+            missingValues.set(val1, new Set(col2.uniqValues))
+        }
+
+        console.timeLog("c", "filled map")
```

Comment:
prettier complains about these - do you want to leave them in for a bit longer and add a comment or take them out?

Reply:
Yeah the plan is to take them out at one point; I only have them in here still because I might fine-tune this a bit more, and it's quite some trouble to add these in.